File: medium/02 Add Two Numbers/_02_Add_Two_Numbers.cpp
// Problem Name: Two Sum
// Difficulty: Medium
// Problem: https://leetcode.com/problems/add-two-numbers/
// Leetcode : https://leetcode.com/u/Shashankpatelc/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* Head = new ListNode(0);
        ListNode* tail = Head;
        int c = 0;

        while (l1 != nullptr || l2 != nullptr || c != 0){
            int d1 = (l1 != nullptr) ? l1->val : 0;
            int d2 = (l2 != nullptr) ? l2->val : 0;

            int sum = d1 + d2 + c;
            int d = sum % 10;
            c = sum / 10;

            ListNode* newNode = new ListNode(d);
            tail->next = newNode;
            tail = tail->next;

            l1 = (l1 != nullptr) ? l1->next : nullptr;
            l2 = (l2 != nullptr) ? l2->next : nullptr;
        }
        ListNode* result = Head->next;
        Head->next = nullptr;
        return result;
    }
};

File: medium/02 Add Two Numbers/_02_Add_Two_Numbers.py
# Problem Name: Two Sum
# Difficulty: Medium
# Problem: https://leetcode.com/problems/add-two-numbers/
# Leetcode : https://leetcode.com/u/Shashankpatelc/

class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        Head = ListNode(0)
        tail = Head
        c = 0

        while l1 is not None or l2 is not None or c != 0:
            d1 = l1.val if l1 is not None else 0
            d2 = l2.val if l2 is not None else 0
            
            sum = d1 + d2 + c
            d = sum % 10
            c = sum // 10

            newNode = ListNode(d)
            tail.next = newNode
            tail = tail.next

            l1 = l1.next if l1 is not None else None
            l2 = l2.next if l2 is not None else None
        
        res = Head.next
        Head.next = None
        return res
        